uber/models.py

```diff
@@ -868,18 +867,22 @@ def single_dept_heads(self, dept=None):
                         .order_by(Attendee.full_name).all())
 
         def match_to_group(self, attendee, group):
-            with c.BADGE_LOCK:
+            with c.ASSIGN_ATTENDEE_TO_GROUP_LOCK:
                 available = [a for a in group.attendees if a.is_unassigned]
                 matching = [a for a in available if a.badge_type == attendee.badge_type]
                 if not available:
                     return 'The last badge for that group has already been assigned by another station'
                 elif not matching:
                     return 'Badge #{} is a {} badge, but {} has no badges of that type'.format(attendee.badge_num, attendee.badge_type_label, group.name)
                 else:
-                    for attr in ['group', 'group_id', 'paid', 'amount_paid', 'ribbon']:
-                        setattr(attendee, attr, getattr(matching[0], attr))
-                    self.delete(matching[0])
```

Comment:
This method of deleting an unassigned group member and assigning a new attendee is broken for groups with numbered badges. The attendee is deleted and the new attendee is assigned before the badge shifting is flushed to the database.

So we have to delete first, then flush, then modify the attendee, then commit.